jerry-core/lit/lit-strings.cpp

```diff
@@ -598,6 +556,183 @@ lit_read_code_point_from_utf8 (const lit_utf8_byte_t *buf_p, /**< buffer with ch
 } /* lit_read_code_point_from_utf8 */
 
 /**
+ * Decodes a unicode code unit from non-empty cesu-8-encoded buffer
+ *
+ * @return number of bytes occupied by code point in the string
+ */
+lit_utf8_size_t
+lit_read_code_unit_from_cesu8 (const lit_utf8_byte_t *buf_p, /**< buffer with characters */
+                               ecma_char_t *code_point) /**< @out: code point */
+{
+  JERRY_ASSERT (buf_p);
+
+  lit_utf8_byte_t c = buf_p[0];
+  if ((c & LIT_UTF8_1_BYTE_MASK) == LIT_UTF8_1_BYTE_MARKER)
+  {
+    *code_point = (lit_code_point_t) (c & LIT_UTF8_LAST_7_BITS_MASK);
+    return 1;
+  }
+
+  lit_code_point_t ret = LIT_UNICODE_CODE_POINT_NULL;
+  ecma_length_t bytes_count = 0;
+  if ((c & LIT_UTF8_2_BYTE_MASK) == LIT_UTF8_2_BYTE_MARKER)
+  {
+    bytes_count = 2;
+    ret = ((lit_code_point_t) (c & LIT_UTF8_LAST_5_BITS_MASK));
+  }
+  else
+  {
+    JERRY_ASSERT ((c & LIT_UTF8_3_BYTE_MASK) == LIT_UTF8_3_BYTE_MARKER);
+    bytes_count = 3;
+    ret = ((lit_code_point_t) (c & LIT_UTF8_LAST_4_BITS_MASK));
+  }
+
+  for (uint32_t i = 1; i < bytes_count; ++i)
```

Comment:
I don't think we need a for here.
Just while (bytes_count-- > 0) and increase buf